src/client/Inode.cc

```diff
@@ -556,3 +564,146 @@ void Inode::set_async_err(int r)
   }
 }
 
+bool Inode::has_recalled_deleg()
+{
+  if (delegations.empty())
+    return false;
+
+  Delegation *deleg = delegations.front();
+
+  return deleg->is_recalled();
+}
+
+void Inode::recall_deleg(bool skip_read)
+{
+  if (delegations.empty())
+    return;
+
+  // Issue any recalls
+  for (xlist<Delegation*>::iterator p = delegations.begin(); !p.end(); ++p) {
+    Delegation *deleg = *p;
+
+    deleg->recall(skip_read);
+  }
+}
+
+bool Inode::delegations_broken(bool skip_read)
+{
+  if (delegations.empty())
+    return true;
+
+  if (skip_read) {
+    Delegation *deleg = delegations.front();
+    if (deleg->get_mode() == CEPH_FILE_MODE_RD)
+	return true;
+  }
+  return false;
+}
+
+void Inode::break_deleg(bool skip_read)
+{
+  recall_deleg(skip_read);
+
+  while (!delegations_broken(skip_read))
+    client->wait_on_list(waitfor_deleg);
+}
+
+/**
+ * set_deleg: request a delegation on an open Fh
+ * @fh: filehandle on which to acquire it
+ * @ro: r/o or r/w delegation?
+ * @cb: delegation recall callback function
+ * @priv: private pointer to be passed to callback
+ *
+ * Attempt to acquire a delegation on an open file handle. If there are no
+ * conflicts and we have the right caps, allocate a new delegation, fill it
+ * out and return 0. Return an error if we can't get one for any reason.
+ */
+int Inode::set_deleg(Fh *fh, bool ro, ceph_deleg_cb_t cb, void *priv)
+{
+  int filemode;
+
+  lsubdout(client->cct, client, 0) << __func__ << ": inode " << *this << dendl;
+
+  // Just say no if we have any recalled delegs still outstanding
+  if (has_recalled_deleg()) {
+    lsubdout(client->cct, client, 0) << __func__ << ": has_recalled_deleg" << dendl;
+    return -EAGAIN;
+  }
+
+  // check vs. currently open files on this inode
+  if (ro) {
+    if (open_count_for_write()) {
+      lsubdout(client->cct, client, 0) << __func__ << ": open for write" << dendl;
+      return -EAGAIN;
+    }
+    filemode = CEPH_FILE_MODE_RD;
+  } else {
+    if (open_count() > 1) {
+      lsubdout(client->cct, client, 0) << __func__ << ": open" << dendl;
+      return -EAGAIN;
+    }
+    filemode = CEPH_FILE_MODE_RDWR;
+  }
+
+  // Do we have the caps we need for this inode?
```

Comment:
sorry to ask this deep down in the code, but broadly, it's surprising we don't "ask for" the deleg toward the MDS;  I guess we don't due that due to greediness?